autofill/autofill-impl/src/main/java/com/duckduckgo/autofill/ui/credential/management/CredentialListSorter.kt

```diff
@@ -28,19 +27,40 @@ interface CredentialListSorter {
 
 @ContributesBinding(FragmentScope::class)
 class CredentialListSorterByTitleAndDomain @Inject constructor(
-    private val domainFormatter: AutofillDomainFormatter
+    private val initialExtractor: InitialExtractor
 ) : CredentialListSorter {
 
     override fun sort(credentials: List<LoginCredentials>): List<LoginCredentials> {
-        return credentials.sortedWith(
-            compareBy(
-                titleSort(),
-                domainSort()
-            )
-        )
+        return credentials.sortedWith(compareBy(titleOrDomainSort()))
     }
 
-    private fun titleSort(): (LoginCredentials) -> Comparable<*>? = { it.domainTitle }
-    private fun domainSort(): (LoginCredentials) -> Comparable<*>? = { domainFormatter.extractDomain(it.domain) }
+    private fun titleOrDomainSort(): (LoginCredentials) -> Comparable<String>? = {
+        val sb = StringBuilder()
 
+        // if we don't have a usable title or domain, return the placeholder so it's sorted first
+        if (!it.canUseDomainInSort() && !it.canUseTitleInSort()) {
+            sb.append("#")
+        }
+
+        // if we can use the title, use it
+        if (it.canUseTitleInSort()) {
+            sb.append(initialExtractor.extractInitialFromTitle(it).toString())
+            sb.append(Char(0))
```

Comment:
Why do we need this `Char(0)`?

Reply:
Sorry, we likely don't. I haven't quite got the sorting/grouping right in all cases and that was an attempt at fixing that. 

I'll ping you for a re-review when I've got all the test cases passing.